### DISCUSSION TITLE
Designate window for new tabs

### DISCUSSION THREAD

**Discussion by BenceSzalai** [l10nelw/winger#4] [Ideas]
This plugin is great help with managing multiple windows, but that concept breaks down a bit when I'm clicking on links in other applications, and as per Firefox's default behaviour it opens in the window that was active last.

It'd be very useful to be able to designate a window for new tabs opened from external apps.

What do you think?

Also what is your stance on contributing to the plugin? Are PRs welcome?

**Comment by l10nelw** [comment]
I'm familiar with the problem but I've always imagined the solution idea as a separate extension, which would catch new tabs from any origin (including other tabs). I definitely want to get it started soon. I take it you'd like to contribute? I welcome PRs.

**Reply by BenceSzalai** [reply]
Well, I don’t have much experience, but that shouldn't be a problem. I've patched few extensions myself before, even though my changes never made upstream, as most was abandoned or badly maintained...

As for catching tabs from external sources, I took a quick look: There is no explicit way in the Bowser Extension JS API afaics.

New tabs can be detected by listening to [`browser.tabs.onCreated`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/onCreated) events. And in order to tell if a tab is coming from outside the browser, [TreeStyleTab checks two things](https://github.com/piroor/treestyletab/blob/15d257d565a7c909f2fd7a151a49ce4f5ef93291/webextensions/background/api-tabs-listener.js#L358):
* `tab.openerTabId` of the new [tab](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/Tab) should be falsey
* none of the [windows](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/windows/Window) returned by `browser.windows.getAll({})` should be `focused` (implemented by TST [here](https://github.com/piroor/treestyletab/blob/15d257d565a7c909f2fd7a151a49ce4f5ef93291/webextensions/background/api-tabs-listener.js#L69-71))

From this point on it is pretty straightforward. We need a way to mark a window as the target for external links and we need to move any tabs detected by the above criteria to the target window. But I assume your extension already has some easy way to move tabs around, so this should be simple. So probably the biggest challenge is to make the UX to designate the default window nice, and to handle the case when the selected window gets closed. I can imagine multiple different approaches to this. A very simple, but still flexible one could be to just add an input field for the default window name: so then if the window exists, new external tabs are moved into it, and if not exist, it can be recreated. This would be sufficient I think for most common use-cases.

What do you think? I can make a proof-of-concept PR, if you like the idea.

As for developing another independent extension, I'm curious why you think this should/could not be part of the Winger extension?

**Reply by l10nelw** [reply]
Thank you for sharing how TST does it; the "how to detect external-origin tabs" research is now practically done! :smile:

>As for developing another independent extension, I'm curious why you think this should/could not be part of the Winger extension?

1. What we're looking at is _automated tab movement_, which I foresee can be more versatile and have more use cases than just catching external-origin tabs. Examples: move all youtube tabs from all windows to a single window, move all non-work tabs out of a work-related window, etc. Catering to one eventually leads to requests to cater to similar use cases. I prefer as little feature creep as possible in extensions. It's best when an extension is limited to doing one or a few things well.

2. The functionality has no real overlap with core Winger. Tab moving is cheap, it comes straight from the `tabs` API. Winger provides active, specific tab movement; the separate extension would provide passive, generalised tab movement; the contrast is clean enough for me to favour separation. Perhaps the extension would need Winger's window names to be more useful... but I could just let it talk to Winger for that data.

**Reply by l10nelw** [reply]
I hope my reasoning is satisfactory, @BenceSzalai. I'm glad you reached out. I'll start building a barebones extension so we can have our solution live as soon as possible. Will keep you posted!

**Reply by BenceSzalai** [reply]
Sure. I was hoping to get this added quickly. I assume rolling out a full flagged independent add-on takes more effort, but I see no point in interfering with your intentions. ;)

Let me know if you need some help or when you have a beta to try out.
